### ISSUE TITLE
String.runInTerminal compatibility issue across Linux distros

### ISSUE THREAD

**Issue by jreus** [supercollider/supercollider#4893]
* SuperCollider version: 3.10.4
* Operating system: Linux Mint 19.3 Tricia
* Other details (Qt version, audio driver, etc.): 

## Steps to reproduce

```supercollider
"ls".runInTerminal; // this command has no effect on Linux Mint
```

## Expected vs. actual behavior

The implementation of String.runInTerminal in exString_linux.sc looks like this:
```

+ String {
	runInTerminal {|shell="sh"|
		("xterm -hold -T 'SuperCollider runInTerminal' -e" + shell + "-c" + this.shellQuote).unixCmd;
	}

	openOS {
		("xdg-open " ++ this.escapeChar($ )).systemCmd
	}
```
The issue is that Linux Mint (and other distros) uses gnome-terminal by default and not xterm. In fact Mint doesn't come packaged with xterm at all. runInTerminal fails silently.

It would be great if this command used the system default terminal out of the box - maybe via a switch statement that checks against some common distros? Or at least should check if xterm exists on the system, or give the option to pass the terminal application choice into the method.

**Comment by brianlheim** [comment]
those all sound like good solutions to me, feel free to submit a patch!

**Comment by dyfer** [comment]
I was thinking about this, too.

I think it would be good to check against a number of commands to see which one is available. After a quick web search I'm thinking about the following:
```
x-terminal-emulator
gnome-terminal
xterm
konsole
terminator
[possibly others?]
```
we could store which one has been found at least in a class variable to avoid checking too often...

One issue is that there might be discrepancies in command line switches to pass as an argument the command to be run in terminal. In that case it might make sense to implement is similarly to how it's done on macOS, where a temporary script is created.

**Comment by brianlheim** [comment]
that seems like a lot of engineering effort for what is, in the end, a really minor feature. can you make it any simpler?

**Comment by joshpar** [comment]
Maybe a classvar that a user can set in startup that says what command to run?

/*
Josh Parmenter
www.realizedsound.net/josh
*/

**Comment by dyfer** [comment]
> that seems like a lot of engineering effort for what is, in the end, a really minor feature. can you make it any simpler?

@brianlheim I don't know. I was thinking about improving this method and I'm sharing my thoughts here. I think that creating a script + checking against a list of various terminal commands is the only way to make it work mostly reliable under various distros. Ultimately this might be up to the person who submits a PR for this/who is willing to work on this. I might do that at some point if nobody does it earlier.

@joshpar I was thinking about that too, but to me this seems to be a random addition (also would we add it to `String`? or `Platform`?). I think "over-engineering" it slightly but making it work OOTB would be better.

In both cases we'd need to take care of any command line switches each individual terminal program requires. I think that building that in would be more user-friendly.

I also agree this is a very minor feature...

**Comment by jreus** [comment]
Even not having it fail silently would be a significant improvement.

**Comment by elgiano** [comment]
I'm looking into this. As you might know already, there doesn't seem to be a single standard way to specify a preferred terminal emulator in linux. So, I would implement a stack of options like:
- a class variable for users to specify which terminal they want (IMO: LinuxPlatform.terminalEmulator)
- if classvar is not set: call a new method, LinuxPlatform.findTerminalEmulator, which:
- tries xdg-terminal, tries x-terminal-emulator, and if everything else fails, goes through an hardcoded list
- when one is found: store it in the classvar
- otherwise: error out nicely

For inspiration, I'm looking at: [](https://gitlab.gnome.org/GNOME/glib/blob/master/gio/gdesktopappinfo.c#L2514)

**Comment by elgiano** [comment]
Here is a sketch, let me know if you think we should go in this direction or not:
```supercollider
~supportedTerminalEmulators = IdentityDictionary[
    // format args are title, shell and command (.shellQuote)
    \rxvt -> "rxvt --hold -T '%' -e % -c %",
    \urxvt -> "urxvt --hold -T '%' -e % -c %",
    \termite -> "termite --hold --title '%' -e \"% -c %\"",
    \konsole -> "konsole --hold --title '%' -e \"% -c %\"",
    \xterm ->  "xterm -hold -T '%' -e \"% -c %\"",
    // can't hold these terminals open via a command-line switch
    // users could append ';read' to their commands
    'gnome-terminal' -> "gnome-terminal -t '%' -- % -c %",
    'mate-terminal' -> "mate-terminal -t '%' -e \"% -c %\"",
    'xfce4-terminal' -> "xfce4-terminal -T '%' -e \"% -c %\"",
    \terminator ->  "terminator -T '%' -e \"% -c %\"",
]

~findTerminalEmulator = {|action|
    {
        var cond = Condition();
        var termFormat = block{|break|
            ~supportedTerminalEmulators.keysValuesDo{|term, formatString|
                "which %".format(term).unixCmd({|exitCode|
                    c.test = exitCode == 0;
                    c.unhang;
                },false);
                c.hang;
                if(c.test){
                    break.(formatString)
                }
            };
            nil
        };
        if(termFormat.notNil){
            ~terminalEmulator = termFormat;
            "Found supported terminal emulator: %".format(termFormat).postln;
            if(action.notNil){action.value(termFormat)};
        }{
            "Couldn't find a supported terminal emulator. Please set LinuxPlatform.terminalEmulator, see Help for details".error;
        }

    }.forkIfNeeded
};


~runInTerminal = {|cmd|
    if(~terminalEmulator.notNil){
        ~terminalEmulator.format('SuperCollider in Terminal','sh',cmd.shellQuote).postln.unixCmd
    }{
        "Looking for a supported terminal emulator...".warn;
        ~findTerminalEmulator.({|term| term.format('SuperCollider in Terminal','sh',cmd.shellQuote.unixCmd)})
    }
}

~runInTerminal.("find . -iname '*.scd'")
// in case your terminal can't be kept open with a command-line 
~runInTerminal.("find . -iname '*.scd'; read")

// explicitly select a terminal
~terminalEmulator = ~supportedTerminalEmulators[\urxvt]
~terminalEmulator = ~supportedTerminalEmulators[\konsole]
~terminalEmulator = ~supportedTerminalEmulators[\termite]
~terminalEmulator = ~supportedTerminalEmulators[\terminator]
~terminalEmulator = ~supportedTerminalEmulators['xfce4-terminal']
~terminalEmulator = ~supportedTerminalEmulators['gnome-terminal']

// or write your own command
~terminalEmulator = "urxvt --hold -T '%' -e % -c %"
```
I feel this stuff should all go in LinuxPlatform, but also let me know if you think it's better to put it somewhere else.

**Comment by dyfer** [comment]
Thanks @elgiano 

I've had a sketch for this that I never finished. For some reason I used `command -v` to check if the command exists but I can't remember why ATM.
- I thought we could check for supported terminals on first run, store the first match in a class variable and use it subsequently (until recompile)
- I haven't thought about the issue with not being able to hold the terminal open. While it seems heavy-handed, we could consider automatically adding `; read` to the command, if that's really the best solution
- I didn't use forking since in my solution the terminal is being selected on the first invocation of `runInTerminal`. Although, since this method is asynchronous anyway, we could maybe consider running in a routine

Here's my implementation for reference. It's UNTESTED (sorry) - that's why I didn't make a PR with it. I'm not saying this is the best approach, but maybe between the two we can choose what's best.

```supercollider
// extString_linux.sc
+ String {
	classvar runInTerminalCmd;
	runInTerminal {|shell="sh"|
		// ("xterm -hold -T 'SuperCollider runInTerminal' -e" + shell + "-c" + this.shellQuote).unixCmd;
		runInTerminalCmd ? {
			var terminalCommands = [
				"x-terminal-emulator -hold -T 'SuperCollider runInTerminal' -e",
				"gnome-terminal -hold -T 'SuperCollider runInTerminal' -e",
				"konsole -hold -T 'SuperCollider runInTerminal' -e",
				"xterm -hold -T 'SuperCollider runInTerminal' -e",
				"terminator -hold -T 'SuperCollider runInTerminal' -e",
			];
			block {|break|
				terminalCommands.do({|cmd|
					if(format("command -v %", cmd).systemCmd == 0) {
						runInTerminalCmd = cmd;
						break.();
					}
				})
			}
			
		}
		if(runInTerminalCmd.notNil) {
			format("% % -c %", terminalCmd, shell, this.shellQuote).unixCmd;
		} {
			// format("% -%: no supported terminal found.", this.class.name, this).warn;
		}
		
	}
	
// ...
}

```

**Comment by elgiano** [comment]
Hey @dyfer

They're in fact almost equivalent IMO.
- let's definitely store the found terminal until recompilation! I'm also
doing it.
- my dictionary implementation makes possible to deal with different ways
to handle arguments, for different emulators, I would stay with it
- I just stumbled in a [stack overflow answer](
https://stackoverflow.com/questions/592620/how-can-i-check-if-a-program-exists-from-a-bash-script)
that says why you should use `command -v` and not `which` :)
- routine or not routine: I like your approach, especially using systemCmd
instead of unixCmd. However (and maybe I'm just overcaring here), forking a
routine could make these checks less blocking by yielding between each
check (in case some user is at the same time busy scheduling patterns).
- organizational: I would make the finding function independent
(findTerminalEmulator), and add it to LinuxPlatform

I will make a PR then, so that we can continue there

**Comment by dyfer** [comment]
> - I just stumbled in a [stack overflow answer]( https://stackoverflow.com/questions/592620/how-can-i-check-if-a-program-exists-from-a-bash-script) that says why you should use `command -v` and not `which` :)

Yes, now I remember that's what I saw too! :)

> - routine or not routine: 

...that is the question.
I agree that it might be safer for other tasks (patterns etc) to have this in a routine. It seems we are not returning anything synchronously anyway.

> - organizational: I would make the finding function independent
(findTerminalEmulator), and add it to LinuxPlatform

I agree that's probably cleaner. I'm just not sure whether adding a method for just one platform is reasonable, esp for such minor functionality in overall. But yeah let's do that for the PR and see if others have any input.

**Comment by madskjeldgaard** [comment]
Was this ever fixed? I am still getting this issue on 3.11.2 on Arch Linux

**Comment by dyfer** [comment]
It was supposed to get into 3.11.2, but looking at the source I can't find it there, must have been omitted by mistake. The fix is definitely in 3.12.0.

**Comment by madskjeldgaard** [comment]
Okay, thanks ! :)
